webui/react/src/services/stream/stream.ts

```diff
@@ -0,0 +1,219 @@
+    this.#syncSent = sync_id;
+  }
+
+  #processMsg(): void {
+    while (this.#pendingMsg.length > 0) {
+      const msg = this.#pendingMsg.shift();
+      if (!msg) break;
+      if (msg['sync_id']) {
+        if (!msg['complete']) {
+          this.#syncStarted = msg['sync_id'];
+          this.#isLoading?.(false);
+        } else {
+          this.#syncComplete = msg['sync_id'];
+          this.#isLoading?.(true);
+        }
+      } else if (this.#syncSent === this.#syncStarted) {
+        // Ignore all messages between when we send a new subscription and when the
+        // sync-start message for that subscription arrives.  These are the online
+        // updates for a subscription we no longer care about.
+
+        forEach(msg, (val, k) => {
+          if (k.includes('_deleted')) {
+            const stream_key = trimEnd(k, '_deleted') as Streamable;
+            const deleted_keys = decode_keys(val as string);
+
+            this.#curSub?.[stream_key].keyCache.delete_msg(deleted_keys);
+            this.#onDelete(stream_key, deleted_keys);
+          } else {
+            this.#curSub?.[StreamEntityMap[k]].keyCache.upsert([val.id], val.seq);
+            this.#onUpsert(msg);
+          }
+        });
+      }
+    }
+  }
+
+  #processSubscription() {
+    if (!this.#curSub && (!this.#subs || this.#subs.length === 0)) return;
+
+    let spec: Record<Streamable, Subscription> | undefined;
+    // Just connected/reconnected
+    if (!this.#syncSent) {
+      // Resend current subscription if current sync not completed
+      if (this.#curSub && (!this.#syncComplete || this.#syncStarted !== this.#syncComplete)) {
+        spec = this.#curSub;
+      } else {
+        spec = this.#subs.shift();
+      }
+      this.#sendSpec(spec);
+    } else if (this.#subs && this.#syncComplete === this.#syncSent) {
+      // for established connection, only send a new sub when current sub is completed
+      spec = this.#subs.shift();
+      this.#sendSpec(spec);
+    }
+  }
+
+  #advance(): void {
+    // We want to shut down
+    if (this.#closedByClient) {
+      if (this.#ws && this.#ws.readyState !== this.#ws.CLOSED) {
+        if (this.#ws.readyState !== this.#ws.CLOSING) {
+          this.#ws.close();
+        }
+      }
+      return;
+    }
+
+    // We got shut down and we wait till timeout finish to reconnect
+    if (this.#ws && this.#ws.readyState === this.#ws.CLOSED) {
+      if (dayjs().isBefore(this.#timeout)) return;
+      this.#ws = undefined;
+    }
+
+    if (!this.#ws) {
+      this.#connect();
+    } else if (this.#ws.readyState === this.#ws.OPEN) {
+      this.#processMsg();
+      this.#processSubscription();
+    }
+  }
+
+  public subscribe(spec: StreamSpec): void {
+    const curSpec = this.#curSub?.[spec.id()];
+    const keyCache = curSpec && !curSpec.spec.equals(spec) ? curSpec.keyCache : new KeyCache();
+
+    this.#subs.push({ ...this.#curSub, [spec.id()]: { keyCache, spec } });
```

Comment:
I'm pretty confused by both of these lines, can you explain?

- `const keyCache = ...`, what does the curSpec have to do with the future subscription, especially if a client can call subscribe() multiple times in a row?
- `subs.push({...this.#curSub, [spec.id()]: { keycache, spec}})`: again, what does `this.#curSub` have to do with anything, why do we need to capture it at all?  Why not just `subs.push(spec)`?

Reply:
- I was trying to reuse the previous keyCache if the new subscription is same as the current subscription, because same subscription should correspond to the same "known", but I think I should have skipped the new subscription if it's the same as the current subscription.
- Suppose `curSub` contains projects and workspaces, then if new subscription contains projects, this would change the projects subscription and keep workspaces subscription same as before.